File: out_of_index.c
#include <stdio.h>

int fixed_length(){
	printf("Fixed length\n");
	int buf_a[5];
	int buf_b[5];
	buf_a[0] = 0;
	// buf_b[5] = 5; // warning

	printf("buf_a address: %p\n", buf_a);
	printf("buf_b address: %p\n", buf_b);

	for (int i = 0; i < 6; i++) {
		buf_b[i] = i;
		printf("buf_b[%d]: %d\n", i, buf_b[i]);
	}
	printf("Done\n");

	printf("buf_a[0]: %d\n", buf_a[0]);

	return 0;
}

int variable_length(){
	printf("Variable length\n");
	// int buflen = 5;
	int buflen = 15;
	int buf_a[buflen];
	int buf_b[buflen];
	buf_a[0] = 0;

	printf("buf_a address: %p\n", buf_a);
	printf("buf_b address: %p\n", buf_b);

	for (int i = 0; i < buflen + 1; i++) {
		buf_b[i] = i;
		printf("buf_b[%d]: %d\n", i, buf_b[i]);
	}
	printf("Done\n");

	printf("buf_a[0]: %d\n", buf_a[0]);

	return 0;
}

int main() {
	// fixed_length();
	variable_length();
	return 0;
}

File: out_of_index.rs
#[allow(dead_code)]
fn fixed_length()->i32 {
	println!("Fixed length");
	let mut buf_a: [i32; 5] = [0; 5]; // [0, 0, 0, 0, 0]
	let mut buf_b: [i32; 5] = [0; 5];
	buf_a[0] = 0;
	// buf_b[5] = 5; // error

	println!("buf_a address: {:p}", &buf_a);
	println!("buf_b address: {:p}", &buf_b);

	for i in 0..6 {
		buf_b[i] = i as i32;
		println!("buf_b[{}]: {}", i, buf_b[i]);
	}
	println!("Done");

	println!("buf_a[0]: {}", buf_a[0]);

	return 0;
}

#[allow(dead_code)]
fn variable_length()->i32 {
	println!("Variable length");
	let buflen = 5;
	
	// Rustの配列は固定長のみ。　下記はエラー 
	// let mut buf_a: [i32; buflen] = [0; buflen];
	// let mut buf_b: [i32; buflen] = [0; buflen];

	// Vecを使う
	let mut buf_a: Vec<i32> = vec![0; buflen];
	let mut buf_b: Vec<i32> = vec![0; buflen];
	buf_a[0] = 0;

	println!("buf_a address: {:p}", &buf_a);
	println!("buf_b address: {:p}", &buf_b);

	for i in 0..buflen+1 {
		buf_b[i] = i as i32;
		println!("buf_b[{}]: {}", i, buf_b[i]);
	}
	println!("Done");

	println!("buf_a[0]: {}", buf_a[0]);

	return 0;
}

fn main() {
	// fixed_length();
	variable_length();
	return;
}
